### PR TITLE
Objective-C Improvements

### PR DESCRIPTION
**Pull request by LunaTheFoxgirl**
I recently did work adding Objective-C improvements to the LDC compiler, this PR aims to add minor improvements to DMD as well, which may then be backported to LDC.

There's some major changes to selector handling in this PR.

First off, selectors will now be automatically generated when a selector attribute has not been provided. The existing selector generation code has been cleaned up to work in a way which makes it easier to expect what will be done.

1. Functions marked as properties will automatically get selectors generated which conforms to Objective-C property syntax eg. `@property int myInt();` will generate `myInt`. `@property void myInt(int);` will generate `setMyInt:`
2. Property functions named with a "is" prefix will have the "is" prefix stripped off for their setter variant.
3. Selector arguments will be named after the parameter _names_ instead of their type, which follows the general expectation one would have of automated selector generation; eg. `void myFunction(int a, int b, int c)` will generate `myFunction:b:c:`

Adding swift stub class support is not really relevant yet until the arm64 backend works, so I'll hold off with that.

### CODE REVIEW

**Comment by rikkimax** [comment]
Nicely done.

It adheres to the requirement that it sticks within the bounds of the objective-c module. As per Walter at yesterdays monthly meeting. So once CI is green, it should be an easy merge.

**Comment by LunaTheFoxgirl** [comment]
This may take a few tries, since I'm doing this on an arm mac building locally is going to be a headache and code completetion is completely broken for some reason when navigating dmd source.

**Comment by LunaTheFoxgirl** [comment]
@dkorpel could the tests be approved again? Fixed the issues that was output in the last test.

**Review by dkorpel — APPROVED** [review]

**Comment by thewilsonator** [comment]
Thanks does this fix any of DMD's objective-C bugs (particularly with respect to selectors?, see the `extern(C/C++/Obj-C)` label.

**Comment by LunaTheFoxgirl** [comment]
> Thanks does this fix any of DMD's objective-C bugs (particularly with respect to selectors?, see the `extern(C/C++/Obj-C)` label.

This PR does not tackle the issues people are reporting, no. Just one I've encountered. If wanted I could look into some of the reported issues in the future.

This PR is mainly about getting selector related feature parity between LDC 1.40 and DMD.

**Comment by thewilsonator** [comment]
Thanks! it would be great if you could add a changelog entry for this.

**Review by thewilsonator — APPROVED** [review]

**Comment by LunaTheFoxgirl** [comment]
Ready to merge? @thewilsonator @dkorpel

File: changelog/dmd.objc-improvements.dd [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    void doSomething(int a, int b, int c) { // `doSomething:b:c:`
+        // ...
+    }
+}
+---
+
+These changes should not break any existing code as the automatic selector generation
+was not present before. And automatic selector generation only applies to extern(Objective-C) methods.
```
Comment (thewilsonator) [inline changelog/dmd.objc-improvements.dd:38]:
```
iff --git a/changelog/dmd.objc-improvements.dd b/changelog/dmd.objc-improvements.dd
index 80e6fe6..735b44b 100644
--- a/changelog/dmd.objc-improvements.dd
+++ b/changelog/dmd.objc-improvements.dd
@@ -35,4 +35,4 @@ class Fox : NSObject {
 ---
 
 These changes should not break any existing code as the automatic selector generation
-was not present before. And automatic selector generation only applies to extern(Objective-C) methods.
\ No newline at end of file
```
Please add a newline to the end of file

**Review by thewilsonator — APPROVED** [review]
otherwise looks good to me

File: changelog/dmd.objc-improvements.dd (hunk 1 above)
Comment (LunaTheFoxgirl) [reply changelog/dmd.objc-improvements.dd:38]: oh seems like dlang bot was a bit quick there

**Comment by thewilsonator** [comment]
I fixed that on master. Don't worry about it.

**Pull request merged**